src/workerd/api/streams/writable-sink-adapter.c++

```diff
@@ -0,0 +1,864 @@
+}
+}  // namespace
+
+WritableStreamSinkKjAdapter::Active::Active(
+    jsg::Lock& js, IoContext& ioContext, jsg::Ref<WritableStream> stream)
+    : ioContext(ioContext),
+      stream(kj::mv(stream)),
+      writer(initWriter(js, this->stream)) {}
+
+WritableStreamSinkKjAdapter::Active::~Active() noexcept(false) {
+  abort(KJ_EXCEPTION(DISCONNECTED, "WritableStreamSinkKjAdapter is canceled."));
+}
+
+void WritableStreamSinkKjAdapter::Active::abort(kj::Exception reason) {
+  if (aborted) return;
+  aborted = true;
+  canceler.cancel(kj::cp(reason));
+  // If the previous read indicated that it was the last read, then
+  // the reader will have already been dropped. We do not need to
+  // cancel it here.
+  ioContext.addTask(ioContext.run([writable = kj::mv(stream), writer = kj::mv(writer),
+                                      exception = kj::cp(reason)](jsg::Lock& js) mutable {
+    auto& ioContext = IoContext::current();
+    auto error = js.exceptionToJsValue(kj::mv(exception));
+    auto promise = writer->abort(js, error.getHandle(js));
+    return ioContext.awaitJs(js, kj::mv(promise));
+  }));
+}
+
+WritableStreamSinkKjAdapter::WritableStreamSinkKjAdapter(
+    jsg::Lock& js, IoContext& ioContext, jsg::Ref<WritableStream> stream)
+    : state(kj::heap<Active>(js, ioContext, kj::mv(stream))),
+      selfRef(kj::rc<WeakRef<WritableStreamSinkKjAdapter>>(
+          kj::Badge<WritableStreamSinkKjAdapter>{}, *this)) {}
+
+WritableStreamSinkKjAdapter::~WritableStreamSinkKjAdapter() noexcept(false) {
+  selfRef->invalidate();
+}
+
+kj::Promise<void> WritableStreamSinkKjAdapter::write(kj::ArrayPtr<const byte> buffer) {
+  auto pieces = kj::arr(buffer);
+  return write(pieces).attach(kj::mv(pieces));
+}
+
+kj::Promise<void> WritableStreamSinkKjAdapter::write(
+    kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) {
+  KJ_SWITCH_ONEOF(state) {
+    KJ_CASE_ONEOF(active, kj::Own<Active>) {
+      KJ_REQUIRE(!active->writePending, "Cannot have multiple concurrent writes.");
+      KJ_IF_SOME(exception, active->pendingAbort) {
+        auto exc = kj::cp(exception);
+        state = kj::mv(exc);
+        return kj::mv(exc);
+      }
+      if (active->closePending) {
+        state.init<Closed>();
+        KJ_FAIL_REQUIRE("Cannot write after close.");
+      }
+      active->writePending = true;
+      return active->canceler
+          .wrap(active->ioContext.run(
+              [self = selfRef.addRef(), writer = active->writer.addRef(), pieces = pieces](
+                  jsg::Lock& js) mutable -> kj::Promise<void> {
+        size_t totalAmount = 0;
+        for (auto piece: pieces) {
+          totalAmount += piece.size();
+        }
+        if (totalAmount == 0) {
+          return kj::READY_NOW;
+        }
+
+        auto backing = jsg::BackingStore::alloc<v8::ArrayBuffer>(js, totalAmount);
+        auto ptr = backing.asArrayPtr();
+        for (auto piece: pieces) {
+          ptr.first(piece.size()).copyFrom(piece);
```

Comment:
It's probably worth a comment that we collapse piecewise writes into one single write, and why we do so. Presumably so that data which our caller wants grouped together for whatever reason is delivered all at once to the JS WritableStream?